File: module-2-agent-loop/solutions/python/01_loop_minimal.py
"""Module 2.1: Minimal agent loop (~50 lines)."""
import sys
import os

sys.path.insert(0, os.path.dirname(__file__))

from dotenv import load_dotenv
import anthropic
from agent_types import Tool

load_dotenv()

client = anthropic.Anthropic()

echo_tool = Tool(
    name="echo",
    description="Echoes back the input message",
    input_schema={
        "type": "object",
        "properties": {"message": {"type": "string", "description": "Message to echo"}},
        "required": ["message"],
    },
    execute=lambda inp: f"Echo: {inp['message']}",
)


def agent_loop(user_message: str, tools: list[Tool]) -> str:
    messages = [{"role": "user", "content": user_message}]
    tool_defs = [
        {"name": t.name, "description": t.description, "input_schema": t.input_schema}
        for t in tools
    ]

    while True:
        response = client.messages.create(
            model="claude-sonnet-4-20250514",
            max_tokens=1024,
            tools=tool_defs,
            messages=messages,
        )

        messages.append({"role": "assistant", "content": response.content})

        if response.stop_reason == "end_turn":
            for block in response.content:
                if block.type == "text":
                    return block.text
            return "(no text response)"

        tool_results = []
        for block in response.content:
            if block.type == "tool_use":
                tool = next((t for t in tools if t.name == block.name), None)
                result = tool.execute(block.input) if tool else f"Error: unknown tool {block.name}"
                tool_results.append({
                    "type": "tool_result",
                    "tool_use_id": block.id,
                    "content": result,
                })
        messages.append({"role": "user", "content": tool_results})


def main():
    result = agent_loop('Use the echo tool to say "Hello from the agent loop!"', [echo_tool])
    print("Agent:", result)


if __name__ == "__main__":
    main()

File: module-2-agent-loop/solutions/01-loop-minimal.ts
/**
 * Module 2.1: Minimal agent loop (~50 lines)
 * send → check stop_reason → execute tools → loop
 */
import "dotenv/config";
import Anthropic from "@anthropic-ai/sdk";
import type { Tool, Message } from "./types.js";

const client = new Anthropic();

// A trivial tool for demonstration
const echoTool: Tool = {
  name: "echo",
  description: "Echoes back the input message",
  input_schema: {
    type: "object" as const,
    properties: { message: { type: "string", description: "Message to echo" } },
    required: ["message"],
  },
  execute: async (input) => `Echo: ${input.message}`,
};

async function agentLoop(userMessage: string, tools: Tool[]) {
  const messages: Message[] = [{ role: "user", content: userMessage }];
  const toolDefs = tools.map(({ name, description, input_schema }) => ({
    name,
    description,
    input_schema: input_schema as Anthropic.Tool["input_schema"],
  }));

  while (true) {
    const response = await client.messages.create({
      model: "claude-sonnet-4-20250514",
      max_tokens: 1024,
      tools: toolDefs,
      messages,
    });

    // Add assistant response to history
    messages.push({ role: "assistant", content: response.content });

    // If no tool use, we're done
    if (response.stop_reason === "end_turn") {
      const text = response.content.find((b) => b.type === "text");
      return text?.type === "text" ? text.text : "(no text response)";
    }

    // Execute each tool call
    const toolResults: Anthropic.ToolResultBlockParam[] = [];
    for (const block of response.content) {
      if (block.type === "tool_use") {
        const tool = tools.find((t) => t.name === block.name);
        const result = tool
          ? await tool.execute(block.input as Record<string, unknown>)
          : `Error: unknown tool ${block.name}`;
        toolResults.push({ type: "tool_result", tool_use_id: block.id, content: result });
      }
    }
    messages.push({ role: "user", content: toolResults });
  }
}

async function main() {
  const result = await agentLoop('Use the echo tool to say "Hello from the agent loop!"', [echoTool]);
  console.log("Agent:", result);
}

main().catch(console.error);
